src/test/java/com/commercetools/sync/sdk2/products/helpers/ProductBatchValidatorTest.java

```diff
@@ -0,0 +1,540 @@
+
+import com.commercetools.api.client.ProjectApiRoot;
+import com.commercetools.api.models.channel.ChannelResourceIdentifierBuilder;
+import com.commercetools.api.models.common.MoneyBuilder;
+import com.commercetools.api.models.common.PriceDraft;
+import com.commercetools.api.models.common.PriceDraftBuilder;
+import com.commercetools.api.models.custom_object.CustomObject;
+import com.commercetools.api.models.customer_group.CustomerGroupResourceIdentifierBuilder;
+import com.commercetools.api.models.product.Attribute;
+import com.commercetools.api.models.product.AttributeBuilder;
+import com.commercetools.api.models.product.ProductDraft;
+import com.commercetools.api.models.product.ProductVariantDraft;
+import com.commercetools.api.models.product.ProductVariantDraftBuilder;
+import com.commercetools.api.models.state.StateResourceIdentifierBuilder;
+import com.commercetools.api.models.tax_category.TaxCategoryResourceIdentifierBuilder;
+import com.commercetools.api.models.type.CustomFieldsDraftBuilder;
+import com.commercetools.api.models.type.FieldContainerBuilder;
+import com.commercetools.api.models.type.TypeResourceIdentifierBuilder;
+import com.commercetools.sync.sdk2.commons.helpers.DefaultCurrencyUnits;
+import com.commercetools.sync.sdk2.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.commercetools.sync.sdk2.products.ProductSyncOptions;
+import com.commercetools.sync.sdk2.products.ProductSyncOptionsBuilder;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class ProductBatchValidatorTest {
+  private List<String> errorCallBackMessages;
+  private ProductSyncOptions syncOptions;
+  private ProductSyncStatistics syncStatistics;
+
+  @BeforeEach
+  void setup() {
+    errorCallBackMessages = new ArrayList<>();
+    final ProjectApiRoot ctpClient = mock(ProjectApiRoot.class);
+    syncOptions =
+        ProductSyncOptionsBuilder.of(ctpClient)
+            .errorCallback(
+                (exception, oldResource, newResource, updateActions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                })
+            .build();
+    syncStatistics = mock(ProductSyncStatistics.class);
+  }
+
+  @Test
+  void getReferencedProductKeys_WithNullAttributes_ShouldReturnEmptySet() {
+    assertThat(getReferencedProductKeys(ProductVariantDraftBuilder.of().build())).isEmpty();
+  }
+
+  @Test
+  void getReferencedProductKeys_WithNoAttributes_ShouldReturnEmptySet() {
+    assertThat(getReferencedProductKeys(ProductVariantDraftBuilder.of().attributes().build()))
```

Comment:
Not sure if this test is the same as the one before? What about adding `emptyList()` of attributes instead?

Reply:
Yes it is the same: https://github.com/commercetools/commercetools-sync-java/blob/89a4a0f194fb0e1666ffc9720c7cbf845f84c640/src/test/java/com/commercetools/sync/products/helpers/ProductBatchValidatorTest.java#L77

Reply:
Hm, wonder if `ProductVariantDraftBuilder.of().attributes().build()` produces the same as `ProductVariantDraftBuilder.of().build()` ? Then it would be better to do `ProductVariantDraftBuilder.of().attributes(emptyList()).build()`?

Reply:
I'm not sure what do you mean here. The old test and the new test are exactly the same. Do you mean that `ProductVariantDraftBuilder` is now internally changed and we should also adapt our tests?